src/Symfony/Bridge/Doctrine/Security/User/EntityUserProvider.php

```diff
@@ -29,6 +29,11 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
  * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ *
+ * @template TUser of UserInterface&PasswordAuthenticatedUserInterface
```

Comment:
```suggestion
 * @template TUser of UserInterface
```

User entities don't have to be password authenticated as far as I know.

Reply:
The class implements PassworUpgraderInterface, with the method:
```
public function upgradePassword(PasswordAuthenticatedUserInterface $user, string $newHashedPassword): void
    {
        $class = $this->getClass();
        if (!$user instanceof $class) {
            throw new UnsupportedUserException(sprintf('Instances of "%s" are not supported.', get_debug_type($user)));
        }

        $repository = $this->getRepository();
        if ($repository instanceof PasswordUpgraderInterface) {
            $repository->upgradePassword($user, $newHashedPassword);
        }
    }
```

So technically it should be password authenticated.

Another way to see it, it has the annotation
```
@template-implements PasswordUpgraderInterface<TUser>
```
and `PasswordUpgraderInterface` has the annotation `@template TUser of PasswordAuthenticatedUserInterface`

Reply:
Hmm, this let's me question whether we should actually want to make any of these interfaces generic.

Implementing `PasswordUpgraderInterface` does not require the user to be password authenticated. As this method implementation demonstrates, we can conditionally upgrade the password (in this case, by decorating the repository).

Reply:
I don't think PasswordUpgraderInterface should be generic, given that the generic type is only used to refine the argument type and this is actually wrong (an upgrader is expected to check supported user classes and throw UnsupportedUserException as there is no `supports` method).

Reply:
> I don't think PasswordUpgraderInterface should be generic, given that the generic type is only used to refine the argument type and this is actually wrong (an upgrader is expected to check supported user classes and throw UnsupportedUserException as there is no `supports` method).

That's exactly the purpose of a generic.

If I have a `FooUgrader` which `implements PasswordUpgraderInterface` and only supports `FooUser`
This won't prevent to check for the argument type and throw an UnsupportedUserException, but this will allow to PHPStan to throw an error If I do:
```
$fooUpgrader->upgradePassord($notFooUser);
```
and statically prevent a runtime exception.

> Hmm, this let's me question whether we should actually want to make any of these interfaces generic. [...] That makes me wonder what advantages such generics gives us.

In the same way, if I have multiples `UserProviderInterface<Foo>`, I can assume that the return type of `loadUserByIdentifier` will be `Foo`. Like values of `Foo[]` are `Foo`...

> As this method implementation demonstrates, we can conditionally upgrade the password (in this case, by decorating the repository).

The repository does work with decoration:
```
if ($repository instanceof PasswordUpgraderInterface) {
        $repository->upgradePassword($user, $newHashedPassword);
}
```
But the param still require a `PasswordAuthenticatedUserInterface $user`

But I understand the conditionally point because you rely on the fact the user do
```
if ($user instance PasswordAuthenticatedUserInterface) {
    $provider->upgradePassword(...);
}
```
Not sure it's classic for PHPStan.

Unless ondrejmirtes has a perfect solution, what can I do is this:
https://github.com/symfony/symfony/pull/50399/commits/b0cdf12e2196d04b3f3f85a0ba57ed0c10315425

Reply:
@VincentLanglet but Symfony will definitely do `$fooUpgrader->upgradePassord($notFooUser);` as Symfony has no way to know whether your upgrader supports that user (see ChainUserProvider for instance).

So this generic type is a lie. your code **must** handle any PasswordAuthenticatedUserInterface as input, throwing an UnsupportedUserException for the ones it does not support. It cannot rely on the caller passing only supported users.

Reply:
> @VincentLanglet but Symfony will definitely do `$fooUpgrader->upgradePassord($notFooUser);` as Symfony has no way to know whether your upgrader supports that user (see ChainUserProvider for instance).
> 
> So this generic type is a lie. your code **must** handle any PasswordAuthenticatedUserInterface as input, throwing an UnsupportedUserException for the ones it does not support. It cannot rely on the caller passing only supported users.

A generic never said you **mustn't** handle any PasswordAuthenticatedUserInterface and throw an exception.
That the same way that prior to php 7, you could write
```
/**
 * @param int $int
 */
function foo($int) {
    if (!is_int($int)) {
        throw new InvalidArgument();
    }
    
    // do things
}
```
The phpdoc was here to **help** the user.

But anyways, that shouldn't be the point here. This PR doesn't add the template to PasswordAuthenticatedUserInterface, and I changed, as requested to.
```
/**
 * @template TUser of UserInterface
 *
 * @template-implements UserProviderInterface<TUser>
 */
class EntityUserProvider implements UserProviderInterface, PasswordUpgraderInterface
```

Reply:
But there is **no** error calling it with another user. That's an expected usage of the interface.

On the other hand, without the generic, phpstan will detect broken implementations of the upgrader that forget to do the check, because it will know that the check is necessary.

Reply:
Can it be discuss somewhere else ? My PR is about the UserProviderInterface, which does have a supportClass method.